Feature/add cell

polikhill opened the pull request.

polikhill committed:
add ProfilesView;

polikhill committed:
Merge branch 'feature/coordinator' into feature/addCell

# Conflicts:
#	OnboardingProject.xcodeproj/project.pbxproj

polikhill committed:
setup IGlistKit;
setup add cell button;

polikhill committed:
add Store, ActionCreator;
setup ProfilesViewModel;
add RxExtensions;

polikhill committed:
setup inserting cell;

polikhill committed:
setup ProfileCell;

polikhill committed:
fix ProfileCell; add middlewares;

polikhill committed:
setup cell editing;

polikhill committed:
fix roomPicker;

polikhill committed:
remove commented code;

aLHF commented on OnboardingProject/Base/Extensions/IGListKitExtensions.swift:
```diff
@@ -0,0 +1,23 @@
+            bundle: nil,
+            for: controller,
+            at: index) as? T else {
+                return nil
```
in such cases it's a programmers error, so let's fatalError here and return non-optional value

aLHF commented on OnboardingProject/Classes/Profiles/View/CollectionCells/AddingCell/AddingCell.swift:
```diff
@@ -0,0 +1,41 @@
+    
+    struct Props: Equatable, Diffable {
+        var diffIdentifier: String {
+            return "\(Date().timeIntervalSince1970)"
```
I believe there is only 1 cell, so we can hardcode the id to some constant

aLHF commented on OnboardingProject/Classes/Profiles/View/CollectionCells/AddingCell/AddingCell.swift:
```diff
@@ -0,0 +1,41 @@
+    override func awakeFromNib() {
+        super.awakeFromNib()
+        plusButton.setTitle(nil, for: .normal)
+        plusButton.setImage(#imageLiteral(resourceName: "plusButton"), for: .normal)
```
💯

aLHF commented on OnboardingProject/Classes/Profiles/View/CollectionCells/ProfileCell/ProfileCell.swift:
```diff
@@ -0,0 +1,125 @@
+  }
+  
+  private func render(props: Props) {
+    nameTextField.text = props.name
```
usually it's a good idea to check if props have changed

```swift 
if props.name != renderedProps?.name {
  nameTextField.text = props.name
}
```

aLHF commented on OnboardingProject/Classes/Profiles/View/CollectionCells/ProfileCell/ProfileCell.swift:
```diff
@@ -0,0 +1,125 @@
+  }
+  
+  private func setupBackground(for state: ValidationState) {
+    var color: UIColor {
```
better idea is to move this logic into makeProps method and then pass backgroundColor in props struct

aLHF commented on OnboardingProject/Classes/Profiles/View/ProfilesController.swift:
```diff
@@ -0,0 +1,93 @@
+import RxSwift
+import RxCocoa
+
+final class ProfilesController: ListBindingSectionController<DiffableBox<ProfilesView.Props>> {
```
let's rename to ProfilesSectionController

aLHF commented on OnboardingProject/Classes/Profiles/View/ProfilesView.swift:
```diff
@@ -0,0 +1,118 @@
+  }()
+  
+  struct Props: Diffable {
+    let addingCellProps: DiffableBox<AddingCell.Props>
```
usually it's a better idea to separate props and diffable protocol.

struct Props {
  let sections: [Section]
}

where Section conforms diffable

aLHF commented on OnboardingProject/Classes/Profiles/ViewModel/Middlewares/ProfilesMiddleware.swift:
```diff
@@ -0,0 +1,58 @@
+      let selectedRooms = store.profiles.map { $0.room }
+      let availableRooms = MeetingRooms.rooms.filter { !selectedRooms.contains($0) }
+      
+      Observable.just(())
```
we don't need just here)

aLHF commented on OnboardingProject/Classes/Profiles/ViewModel/Middlewares/ProfilesMiddleware.swift:
```diff
@@ -0,0 +1,58 @@
+extension ProfilesList {
+  static func makeAddNewCellMiddleware() -> Store.Middleware {
+    let disposeBag = DisposeBag()
+    var i = 0
```
let's give it a more meaningful name

aLHF commented on OnboardingProject/Classes/Profiles/ViewModel/Middlewares/ProfilesMiddleware.swift:
```diff
@@ -0,0 +1,58 @@
+        return
+      }
+      
+      Observable.just(())
```
no need for just

aLHF commented on OnboardingProject/Classes/Profiles/ViewModel/ProfilesProps.swift:
```diff
@@ -0,0 +1,58 @@
+  }
+  
+  static private func makeProfileCellProps(
+    from profiles: [ProfileInfo],
```
it's okay to pass state to helper functions

polikhill committed:
add Tagged;
fix ListCollectionContext extension;
move creating color logic to ProfileProps from ProfileCell;
check if props are equal to rendered props in ProfileCell;

polikhill committed:
separate ProfileViewProps and Diffable protocol;

polikhill committed:
remove Rx from Middleware;
fix ProfileProps;

polikhill committed:
fix cells reuse;

aLHF commented on OnboardingProject/Models/Sections.swift:
```diff
@@ -0,0 +1,23 @@
+  let profileCellProps: [DiffableBox<ProfileCell.Props>]
+  
+  var diffIdentifier: String {
+    return UUID().uuidString
```
?

aLHF commented on OnboardingProject/Classes/Profiles/View/ProfilesSectionController.swift:
```diff
@@ -54,15 +54,18 @@ extension ProfilesSectionController: ListBindingSectionControllerDataSource {
         .bind(to: roomSubject)
         .disposed(by: cell.disposeBag)
       
+      print("=== pr index \(index)")
```
let's remove prints from PR
